### ISSUE TITLE
Can caching in `h5py.File(..., "r")` be completely disabled?

### ISSUE THREAD

**Issue by pzelasko** [h5py/h5py#2023]
Summary of the h5py configuration
---------------------------------

h5py    3.6.0
HDF5    1.12.1
Python  3.8.12 (default, Oct 12 2021, 13:49:34)
[GCC 7.5.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.20.3
cython (built with) 0.29.24
numpy (built against) 1.17.3
HDF5 (built against) 1.12.1

I would like to completely disable all caching in HDF5/h5py, as for some use-cases it doesn't make sense (deep learning training on very large data where each sample is read only once). I am using `h5py` as dependency in my library `lhotse` and some users are reporting they see growing memory consumption, which I tracked down to our use of `h5py` (details are in this issue: https://github.com/lhotse-speech/lhotse/issues/518).

I tried to find the instructions for disabling caches and got this, but none of the options had any effect on my observed (growing) memory usage with time: https://docs.h5py.org/en/stable/high/file.html?highlight=cache#chunk-cache

**Comment by takluyver** [comment]
HDF5's caches are quite small by default, as far as I know. So if you're seeing memory usage continually growing through many GB, that's probably a bug somewhere rather than caching. There are some memory leaks already tracked, e.g. #1980, #2010 .

**Comment by pzelasko** [comment]
I see. Just to make sure: in our scenarios, there are usually multiple HDF5 files open (20-80), each by multiple data loading processes (4-8), again multiplied by the number of GPUs (1-4). And every training sample is a separate HDF5 dataset, since they are speech utterances and have different lengths. I thought that since caches are per dataset, that could naturally lead to memory explosion.

**Comment by takluyver** [comment]
It should be OK so long as you don't keep the datasets open after reading them. In Python this means keeping a reference to the dataset object. So something like this should be OK:

```python
f = h5py.File('foo.h5', 'r')
# Only hold one dataset object at a time
for name in dset_names:
    dset = f[name]
    process(dset[:])
```

But something like this could cause a problem:

```python
f = h5py.File('foo.h5', 'r')
# Keep hold of all the dataset objects
dsets = [f[name] for name in dset_names]
for dset in dsets:
    process(dset[:])
```

**Comment by pzelasko** [comment]
Thanks for getting back to me! I don't think I'm holding a reference to dataset objects anywhere, although I'm keeping the file itself open in a global cache. This is the usage:

https://github.com/lhotse-speech/lhotse/blob/912259f09407900fd0c02b9c8f04e20cb4d14784/lhotse/features/io.py#L480-L491

So far the only thing that seems to help is periodically closing and re-opening the HDF5 file -- it limits the memory usage, and the longer the file is open, the more memory seems to be consumed (and never freed).

**Comment by antoinebrl** [comment]
Hi @pzelasko. The cache has a default size of 1MB. Have a look at [`rdcc_nbytes`](https://docs.h5py.org/en/latest/high/file.html#h5py.File) if you want to reduce its size. Hope this solves your issue or helps identify the cause of the increasing memory footprint.

**Comment by pzelasko** [comment]
Thanks @antoinebrl, I tried these settings but unfortunately nothing ever helped. I believe the issue could be related to the usage of `np.void()` type that wraps binary data, as I don't think I have ever experienced these issues with int and float types.

**Comment by dyhan316** [comment]
@pzelasko 

Hello, have you been able to resolve the issue?

**Comment by pzelasko** [comment]
No.

**Comment by dyhan316** [comment]
Oh I see.. Thank you for letting me know!

**Comment by huberb** [comment]
Opened a related issue. I also see memory usage increase when closing filehandles after each read:

https://github.com/h5py/h5py/issues/2553
